test.sh

```diff
@@ -44,7 +44,7 @@ assert_is_subcommit_of () {
 rest_of_tree () (
 	export GIT_INDEX_FILE=.git/index.tmp
 	git read-tree $1
-	git rm --cached -r $(test $2 && echo $2 || echo path/to/sub/) -q &>/dev/null
+	git rm --cached -r $(test $2 && echo $2 || echo path/to/sub/) -q >/dev/null 2>&1
```

Comment:
Running with the Ubuntu default shell "dash" caught it, it just took a while to debug it and understand what was going on!  The effect of this was that this "git rm" ran in the background in parallel with the "git write-tree" immediately after and therefore gave inconsistent results.